include/mumble/Base64.hpp

```diff
@@ -3,30 +3,134 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
+// Inspired by https://renenyffenegger.ch/notes/development/Base64/Encoding-and-decoding-base-64-with-cpp.
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, std::numeric_limits< uint8_t >::max() + 1 > table = {
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55,
+			56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12,
+			13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64, 64, 26, 27, 28, 29, 30, 31, 32,
+			33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
+
+		if (in.empty() || in.size() % 4) {
+			return 0;
+		}
+
+		size_t size = in.size() / 4 * 3;
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2 : 1;
+		}
+
+		if (!out.data()) {
+			return size;
+		}
+
+		if (out.size() < size) {
+			return 0;
+		}
+
+		ByteView data(reinterpret_cast< uint8_t * >(out.data()), out.size());
+
+		while (!in.empty()) {
+			data[0] = (table[in[0]] << 2) + ((table[in[1]] & 0x30) >> 4);
+
+			if (in[2] == '=') {
+				break;
+			}
+
+			data[1] = ((table[in[1]] & 0x0f) << 4) + ((table[in[2]] & 0x3c) >> 2);
+
+			if (in[3] == '=') {
+				break;
+			}
 
-	Base64();
-	virtual ~Base64();
+			data[2] = ((table[in[2]] & 0x03) << 6) + table[in[3]];
 
-	virtual explicit operator bool();
+			in   = in.subspan(4);
+			data = data.subspan(std::min(size_t{ 3 }, data.size()));
```

Comment:
What is the advantage of using this sub-spabn approach to a more traditional for-loop? Technically, this subspan view requires constant creation of temporary objects and also a for-loop makes it more obvious that this is not an indeterminant loop, but one that has a precise iteration count.

Reply:
In this specific case the sub-span is useful because it scales the size automatically while advancing the pointer.

Reply:
On a quick glance we don't seem to access the size inside the loop, do we? :eyes:

Reply:
Sorry, we do it in `encode()`. In `decode()`'s loop we could indeed avoid reassigning the span and simply advancing the pointer.